File: core-ext/src/main/java/com/github/jonathanmerritt/rxassetmanager/core/ext/IsRxAssetManager.java
/*
 *     Copyright 2018 Jonathan Merritt 11R00TT00R11@GMAIL.COM
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 */

package com.github.jonathanmerritt.rxassetmanager.core.ext;

import android.content.res.AssetFileDescriptor;
import android.content.res.XmlResourceParser;
import io.reactivex.Flowable;
import io.reactivex.Maybe;
import java.io.File;
import java.io.InputStream;

import static android.content.res.AssetManager.ACCESS_STREAMING;

public interface IsRxAssetManager extends com.github.jonathanmerritt.rxassetmanager.core.IsRxAssetManager {

  Maybe<String> openString(String fileName, int accessMode);

  default Maybe<String> openString(String fileName) {
    return openString(fileName, ACCESS_STREAMING);
  }

  Maybe<byte[]> openBytes(String fileName, int accessMode);

  default Maybe<byte[]> openBytes(String fileName) {
    return openBytes(fileName, ACCESS_STREAMING);
  }

  Maybe<File> openSave(String fileName, int accessMode, String saveFolder);

  default Maybe<File> openSave(String fileName, String saveFolder) {
    return openSave(fileName, ACCESS_STREAMING, saveFolder);
  }

  Flowable<String> listAll(String folderName);

  Flowable<InputStream> listOpen(String folderName, int accessMode, boolean listAll);

  default Flowable<InputStream> listOpen(String folderName, boolean listAll) {
    return listOpen(folderName, ACCESS_STREAMING, listAll);
  }

  default Flowable<InputStream> listOpen(String folderName, int accessMode) {
    return listOpen(folderName, accessMode, false);
  }

  default Flowable<InputStream> listOpen(String folderName) {
    return listOpen(folderName, false);
  }

  Flowable<String> listOpenString(String folderName, int accessMode, boolean listAll);

  default Flowable<String> listOpenString(String folderName, boolean listAll) {
    return listOpenString(folderName, ACCESS_STREAMING, listAll);
  }

  default Flowable<String> listOpenString(String folderName, int accessMode) {
    return listOpenString(folderName, accessMode, false);
  }

  default Flowable<String> listOpenString(String folderName) {
    return listOpenString(folderName, false);
  }

  Flowable<byte[]> listOpenBytes(String folderName, int accessMode, boolean listAll);

  default Flowable<byte[]> listOpenBytes(String folderName, boolean listAll) {
    return listOpenBytes(folderName, ACCESS_STREAMING, listAll);
  }

  default Flowable<byte[]> listOpenBytes(String folderName, int accessMode) {
    return listOpenBytes(folderName, accessMode, false);
  }

  default Flowable<byte[]> listOpenBytes(String folderName) {
    return listOpenBytes(folderName, false);
  }

  Flowable<File> listOpenSave(String folderName, int accessMode, String saveFolder, boolean listAll);

  default Flowable<File> listOpenSave(String folderName, String saveFolder, boolean listAll) {
    return listOpenSave(folderName, ACCESS_STREAMING, saveFolder, listAll);
  }

  default Flowable<File> listOpenSave(String folderName, int accessMode, String saveFolder) {
    return listOpenSave(folderName, accessMode, saveFolder, false);
  }

  default Flowable<File> listOpenSave(String folderName, String saveFolder) {
    return listOpenSave(folderName, saveFolder, false);
  }

  Flowable<AssetFileDescriptor> listOpenFd(String folderName, boolean listAll);

  default Flowable<AssetFileDescriptor> listOpenFd(String folderName) {
    return listOpenFd(folderName, false);
  }

  Flowable<AssetFileDescriptor> listOpenNonAssetFd(int cookie, String folderName, boolean listAll);

  default Flowable<AssetFileDescriptor> listOpenNonAssetFd(String folderName, boolean listAll) {
    return listOpenNonAssetFd(0, folderName, listAll);
  }

  default Flowable<AssetFileDescriptor> listOpenNonAssetFd(int cookie, String folderName) {
    return listOpenNonAssetFd(cookie, folderName, false);
  }

  default Flowable<AssetFileDescriptor> listOpenNonAssetFd(String folderName) {
    return listOpenNonAssetFd(folderName, false);
  }

  Flowable<XmlResourceParser> listOpenXmlResourceParser(int cookie, String folderName, boolean listAll);

  default Flowable<XmlResourceParser> listOpenXmlResourceParser(String folderName, boolean listAll) {
    return listOpenXmlResourceParser(0, folderName, listAll);
  }

  default Flowable<XmlResourceParser> listOpenXmlResourceParser(int cookie, String folderName) {
    return listOpenXmlResourceParser(cookie, folderName, false);
  }

  default Flowable<XmlResourceParser> listOpenXmlResourceParser(String folderName) {
    return listOpenXmlResourceParser(folderName, false);
  }
}

File: core-ext/src/main/kotlin/com/github/jonathanmerritt/rxassetmanager/core/ext/IsRxAssetManager.kt
/*
 *     Copyright 2018 Jonathan Merritt 11R00TT00R11@GMAIL.COM
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 */

package com.github.jonathanmerritt.rxassetmanager.core.ext

import android.content.res.AssetFileDescriptor
import android.content.res.XmlResourceParser
import android.graphics.Bitmap
import android.graphics.Typeface
import com.github.jonathanmerritt.rxassetmanager.core.IsRxAssetManager.Companion.COOKIE
import com.github.jonathanmerritt.rxassetmanager.core.IsRxAssetManager.Companion.MODE
import com.github.jonathanmerritt.rxassetmanager.core.IsRxAssetManager.Companion.PATH
import io.reactivex.Flowable
import io.reactivex.Flowable.empty
import io.reactivex.Maybe
import java.io.File
import java.io.InputStream
import com.github.jonathanmerritt.rxassetmanager.core.IsRxAssetManager as isRxAssetManager

/**
 * Reactive asset manager extended interface.
 *
 * This class outlines the extended api for the library.
 */
interface IsRxAssetManager : isRxAssetManager {
  /**
   * Companion object used for default parameter values.
   */
  companion object {
    /**
     * Default list all = false.
     */
    const val ALL = false
  }

  /**
   * Defaults to an empty maybe.
   */
  fun openString(path: String, mode: Int = MODE): Maybe<String> = Maybe.empty()

  /**
   * Defaults to an empty maybe.
   */
  fun openStringPair(path: String, mode: Int = MODE): Maybe<Pair<String, String>> = Maybe.empty()

  /**
   * Defaults to an empty maybe.
   */
  fun openBytes(path: String, mode: Int = MODE): Maybe<ByteArray> = Maybe.empty()

  /**
   * Defaults to an empty maybe.
   */
  fun openBytesPair(path: String, mode: Int = MODE): Maybe<Pair<String, ByteArray>> = Maybe.empty()

  /**
   * Defaults to an empty maybe.
   */
  fun openSave(path: String, mode: Int = MODE, to: String): Maybe<File> = Maybe.empty()

  /**
   * Defaults to an empty maybe.
   */
  fun openSavePair(path: String, mode: Int = MODE, to: String): Maybe<Pair<String, File>> = Maybe.empty()

  /**
   * Defaults to an empty maybe.
   */
  fun openBitmap(path: String, mode: Int = MODE): Maybe<Bitmap> = Maybe.empty()

  /**
   * Defaults to an empty maybe.
   */
  fun openBitmapPair(path: String, mode: Int = MODE): Maybe<Pair<String, Bitmap>> = Maybe.empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listAll(path: String = PATH, sorting: Sorting = Normal): Flowable<String> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpen(path: String = PATH, mode: Int = MODE, all: Boolean = ALL): Flowable<InputStream> =
      empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenPair(path: String = PATH, mode: Int = MODE, all: Boolean = ALL):
      Flowable<Pair<String, InputStream>> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenString(path: String = PATH, mode: Int = MODE, all: Boolean = ALL): Flowable<String> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenStringPair(path: String = PATH, mode: Int = MODE, all: Boolean = ALL):
      Flowable<Pair<String, String>> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenBytes(path: String = PATH, mode: Int = MODE, all: Boolean = ALL): Flowable<ByteArray> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenBytesPair(path: String = PATH, mode: Int = MODE, all: Boolean = ALL):
      Flowable<Pair<String, ByteArray>> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenSave(path: String = PATH, mode: Int = MODE, to: String, all: Boolean = ALL): Flowable<File> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenSavePair(path: String = PATH, mode: Int = MODE, to: String, all: Boolean = ALL):
      Flowable<Pair<String, File>> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenBitmap(path: String = PATH, mode: Int = MODE, all: Boolean = ALL): Flowable<Bitmap> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenBitmapPair(path: String = PATH, mode: Int = MODE, all: Boolean = ALL):
      Flowable<Pair<String, Bitmap>> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenTypeface(path: String = PATH, all: Boolean = ALL): Flowable<Typeface> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenTypefacePair(path: String = PATH, all: Boolean = ALL): Flowable<Pair<String, Typeface>> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenFd(path: String = PATH, all: Boolean = ALL): Flowable<AssetFileDescriptor> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenFdPair(path: String = PATH, all: Boolean = ALL): Flowable<Pair<String, AssetFileDescriptor>> =
      empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenNonAssetFd(cookie: Int = COOKIE, path: String = PATH,
      all: Boolean = ALL): Flowable<AssetFileDescriptor> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenNonAssetFdPair(cookie: Int = COOKIE, path: String = PATH, all: Boolean = ALL):
      Flowable<Pair<String, AssetFileDescriptor>> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenXmlResourceParser(cookie: Int = COOKIE, path: String = PATH, all: Boolean = ALL):
      Flowable<XmlResourceParser> = empty()

  /**
   * Defaults to an empty flowable.
   */
  fun listOpenXmlResourceParserPair(cookie: Int = COOKIE, path: String = PATH, all: Boolean = ALL):
      Flowable<Pair<String, XmlResourceParser>> = empty()
}
